Makefile

```diff
@@ -125,17 +125,13 @@ build-container: check-sources ## Builds components required for cilium-agent co
 $(SUBDIRS): force ## Execute default make target(make all) for the provided subdirectory.
 	@ $(MAKE) $(SUBMAKEOPTS) -C $@ all
 
-# If the developer provides TESTPKGS to filter the set of packages to use for testing, filter that to only packages with privileged tests.
-# The period at EOL ensures that if TESTPKGS becomes empty, we can still pass some paths to 'xargs dirname' to avoid errors.
-PRIV_TEST_PKGS_FILTER := $(shell for pkg in $(TESTPKGS); do echo $$pkg; done | xargs grep --include='*.go' -ril 'go:build [^!]*privileged_tests') .
-PRIV_TEST_PKGS_EVAL := $(shell echo $(PRIV_TEST_PKGS_FILTER) | xargs dirname | sort | uniq | grep -Ev '^\.$$')
-PRIV_TEST_PKGS ?= $(PRIV_TEST_PKGS_EVAL)
-tests-privileged: GO_TAGS_FLAGS+=privileged_tests ## Run integration-tests for Cilium that requires elevated privileges.
-tests-privileged:
+tests-privileged: ## Run integration tests for Cilium that require elevated privileges.
 	$(MAKE) init-coverage
-	for pkg in $(patsubst %,github.com/cilium/cilium/%,$(PRIV_TEST_PKGS)); do \
+	for pkg in $(patsubst %,github.com/cilium/cilium/%,$(TESTPKGS)); do \
 		>&2 $(ECHO_TEST) $$pkg; \
-		PATH=$(PATH):$(ROOT_DIR)/bpf $(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
+		PATH=$(PATH):$(ROOT_DIR)/bpf \
+		PRIVILEGED_TESTS=true \
+		$(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
```

Comment:
Couple of possible areas to explore:
- Ensure there is a way to skip unprivileged tests, so that jenkins test doesn't need to run those additional tests that are already run elsewhere
- Move privileged tests out of the jenkins CI into a github action, so we don't extend the jenkins test job runtimes but instead the privileged tests run in parallel with other jobs

Reply:
@joestringer Thanks for taking a look! I had all the points you raised in mind while working on this.

First,

> Does this mean that the `tests-privileged` now runs all tests, not just the privileged ones?

Yes! And that's actually fine, because unit ('unprivileged') tests are actually very fast to run, as they should be. The most expensive part of running our test suite is building test binaries. Being able to run all unit, privileged and integration tests in one go is one of the end goals of these changes.

With build cache (4.38s):
```
~ time go test ./...
...
go test ./...  29.60s user 4.38s system 763% cpu 4.450 total
```

Without build cache:
```
~ go clean -testcache ./...
~ time go test ./...
go test ./...  245.62s user 25.07s system 835% cpu 32.383 total
```

This is on a Ryzen 3700X (16 threads) desktop CPU with fast memory, with all source files in OS cache. Even on this machine, build time accounts for 87% of total test suite runtime. I would expect this number to increase even further on cloud VMs with noisy neighbours and slower i/o.

IMO the only proper way of skipping unprivileged tests would be by consistenly naming test functions, e.g. `func Test(Un)PrivilegedFoo()` so they can be filtered out by `go test -run=TestPrivileged`. This is not a realistic goal given the amount of testing frameworks in use throughout the codebase that all have their own quirky way of doing things.

In conclusion: always running unit tests is fine and barely amounts to a rounding error on a total privileged test runtime of 2.5+ hours.

---

Second, regarding Jenkins/Travis/GHA, ...

For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

Note that we need nested virt for privileged tests, so we'd need self-hosted GHA runners first.

We'll get there, this is just a first attempt at moving the needle. :slightly_smiling_face:

Reply:
:+1: fair enough. I'm always worried about "just a couple more minutes" on the overall build time explicitly because we mostly just tolerate the test length and we don't grapple with how to make it shorter. So any time a PR like this comes up, the easiest solution is just to tolerate making our dev cycle even longer.

That said, in general we're currently reviewing the jenkins tests for which ones we can convert to other forms of testing / delete so we should be able to make up for this additional test run time pretty soon:tm:. And yeah it's not really fair to hold up this particular PR if this is a general problem with the testing code in Cilium.

> For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

:+1: I think the only reason we do it on Jenkins today is to be able to run the tests in a privileged environment. Wherever we can get that, we can run it there.